Mount propagation in kubelet

jsafrane opened the pull request:
Together with #45724 it implements mount propagation as proposed in https://github.com/kubernetes/community/pull/659

There is:

- New alpha annotation that allows user to explicitly set propagation mode for each `VolumeMount` in pod containers (to be replaced with real `VolumeMount.Propagation` field during beta) + validation + tests. "Private" is the default one (= no change to existing pods).

  I know about proposal for real API fields for alpha feature in https://docs.google.com/document/d/1wuoSqHkeT51mQQ7dIFhUKrdi3-1wbKrNWeIL4cKb9zU/edit, but it seems it's not implemented yet. It would save me quite lot of code and ugly annotation.

- Updated CRI API to transport chosen propagation to Docker.

- New `kubelet --experimental-mount-propagation` option to enable the previous bullet without modifying types.go (worked around with changing `KubeletDeps`... not nice, but it's better than adding a parameter to `NewMainKubelet` and removing it in the next release...)

```release-note
kubelet has alpha support for mount propagation. It is disabled by default and it is there for testing only. This feature may be redesigned or even removed in a future release.
```

@derekwaynecarr @dchen1107 @kubernetes/sig-node-pr-reviews

jsafrane commented:
And I promise I'll write an e2e or node conformance test when #45724 lands and if there is a way how to add `--test-mount-propagation` to kubelet under test. Any suggestion/precedent?

derekwaynecarr commented:
@jsafrane -- if you wanted to use this in node e2e test, you can control how kubelet is launched here: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/services/kubelet.go#L130

derekwaynecarr commented on pkg/kubelet/apis/cri/v1alpha1/api.pb.go:
```diff
@@ -161,6 +161,30 @@ func (x Protocol) String() string {
 }
 func (Protocol) EnumDescriptor() ([]byte, []int) { return fileDescriptorApi, []int{0} }
 
+type MountPropagation int32
```
cc @yujuhong as primary reviewer for CRI changes.

derekwaynecarr commented on pkg/kubelet/dockershim/helpers.go:
```diff
@@ -122,28 +122,39 @@ func extractLabels(input map[string]string) (map[string]string, map[string]strin
+// 'ro', if the path is read only
+// 'Z', if the volume requires SELinux relabeling
+// propagation mode such as 'rslave'
 func generateMountBindings(mounts []*runtimeapi.Mount) (result []string) {
```
it would be nice if we had a unit test for this function.

jsafrane replied:
unit test added

derekwaynecarr commented on pkg/kubelet/kubelet.go:
```diff
@@ -232,6 +232,15 @@ type KubeletDeps struct {
 	Writer             kubeio.Writer
 	VolumePlugins      []volume.VolumePlugin
 	TLSOptions         *server.TLSOptions
+	MountOptions       KubeletMountOptions
```
i thought this would be a pointer.

jsafrane replied:
pointer added

derekwaynecarr commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -117,6 +117,12 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 	glog.V(3).Infof("container: %v/%v/%v podIP: %q creating hosts mount: %v", pod.Namespace, pod.Name, container.Name, podIP, mountEtcHostsFile)
 	mounts := []kubecontainer.Mount{}
+
+	propagations, err := v1helper.GetPodPropagation(pod.Annotations)
```
i am wondering if the kubelet should reject pods that use this annotation during admission if the feature is not enabled.

i know we said we would deprecate the annotation, but i dont always trust our future self.

jsafrane replied:
Doesn't admission controller run on API server? Or kubelet has it's own?

jsafrane replied:
`mountPropagationAdmissionHandler` added

jsafrane commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

Random-Liu commented:
Reassigning to @dchen1107, because she should have better idea on this.

gurvindersingh commented:
@jsafrane I am not sure as which one will get merge to master for supporting mount propagation. But I am wondering in this PR case the POD author has to add this annotation himself/herself. Wondering if there is a way to auto annotate PODs or it would be an idea to pick the annotation from PVC like here done for GID https://github.com/kubernetes/kubernetes/blob/1e1040b524dda1c97dad4faab1910972b71a5c79/pkg/volume/util/volumehelper/volumehelper.go#L42 as then if there is an annotation on PVC for propagation mode then PODs using those PVC will use that, POD can override the mode if they prefer by adding annotation themselves if that's preferred.

I would like to have a way where I, as a cluster admin, can control the propagation mode for certain PVC which requires mount propagation e.g. backed by distributed file systems.

yujuhong commented:
@jsafrane @derekwaynecarr maybe I missed some discussions on this topic, but I think the change to the proposal (https://github.com/kubernetes/community/pull/659) should get reviewed first before we jump into the reviewing the code.

Is this slated for 1.7?

jsafrane commented:
@yujuhong we have mount containers on sig-storage 1.7 list. The agreed proposal won't work (and it was not implemented anyway), I'm trying to catch 1.7 with my update in https://github.com/kubernetes/community/pull/659

jsafrane commented:
@gurvindersingh, only HostPath volumes can be slave or shared and the mount propagation is property of the *destination* directory where the HostPath volume is mounted, not the source volume (i.e. PV/PVC). So it does not make much sense to add propagation to PVs/PVCs.

Do you have any specific use case in mind?

gurvindersingh commented:
@jsafrane PV in this case I mentioned above is hostpath and PVC is the for the same PV. As for security reason we don't allow users to use hostPath directly but a cluster admin creates PV for hostpath and then users can use them using PVC (which are backed by hostPath PV). In my use case I have distributed file system like "GlusterFS" mounted on all worker nodes and then have hostpath PV created from certain directory path depending upon the project. On those PVs, I would like to have "rslave" mode attached as when POD using them with PVCs can get mount/remount events propagated in the case of issue with storage/network layer. So wondering if it make sense to add the mount propagation annotation on PV/PVC to be automated by cluster admin on certain PVs/PVCs.

dims commented:
/assign @Random-Liu 
/unassign

jsafrane commented:
rebased and fixed verification test

jsafrane commented:
@gurvindersingh, sorry for late reply. Initially I thought that we should start with "private" as default and then move to "rslave". Now I see a real use case for "rslave" as default, so let's try to default to "rslave" from start and fail early. "rslave" as default was already envisioned in original [propagation.md](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/propagation.md)

I updated https://github.com/kubernetes/community/pull/659, I'll update (and rebase) this PR shortly.

jsafrane commented:
- rebased
- made rslave default when --test-mount-propagation is set (otherwise defaults to private as in previous Kubernetes releases)

jsafrane commented:
/retest

runcom commented:
/cc @mrunalp @rhvgoyal

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -80,6 +80,10 @@ type KubeletFlags struct {
+
+	// Experimental flag that enables shared/slave propagation for mounts in
+	// pods. It may be redesigned or even removed in future releases!
+	MountPropagation bool
```
@mtaufen what's the new policy for experimental flags?

mtaufen replied:
We talked about this specific flag in sig-node and agreed it was ok for now. Can you please name it `ExperimentalMountPropagation` though (and the flag `--experimental-mount-propagation`), just to make it really clear that it's experimental?

We don't want to keep adding experimental stuff to the Kubelet's versioned configuration types at the moment, so the only option is to add flags for experimental things. At some point (hopefully soon) apimachinery will have field gates (https://docs.google.com/document/d/1wuoSqHkeT51mQQ7dIFhUKrdi3-1wbKrNWeIL4cKb9zU/edit#), and we can probably use that to include experimental things in the versioned configuration type (without having to rev the version), rather than as flags.

jsafrane replied:
renamed to `--experimental-mount-propagation` (we agreed on `--test-*` on the meeting though...)

ddysher commented on cmd/kubelet/app/options/options.go:
```diff
@@ -142,6 +146,12 @@ func (f *KubeletFlags) AddFlags(fs *pflag.FlagSet) {
+	// test-mount-propagation will be removed in a future release either because
+	// mount propagation works and will be enabled by default or does not work
+	// and we need to get back to the drawing board.
+	fs.BoolVar(&f.MountPropagation, "test-mount-propagation", f.MountPropagation, "Enables shared/slave mount propagation from pods. For testing only.")
```
should we follow other flag convention and use `experimental` instead?

jsafrane replied:
It has been decided on sig-node meeting that `--test` is the right prefix here.

jsafrane replied:
and now the decision changed, renamed to `--experimental-mount-propagation`

ddysher replied:
thanks, this is clearer IMO.

ddysher commented on pkg/api/helper/helpers.go:
```diff
@@ -602,3 +603,23 @@ func StorageNodeAffinityToAlphaAnnotation(annotations map[string]string, affinit
+
+type MountPropagationMap map[string]map[string]api.MountPropagation
+
+func GetPodPropagation(podAnnotations map[string]string) (MountPropagationMap, error) {
```
Add godoc for exported function?  I think some lint tools will complain about this...

jsafrane replied:
good catch, fixed.

jsafrane commented:
- fixed review remarks
- reworked the e2e test to run on Docker >= 1.13. Feel free to suggest a better way. Am I correct that node e2e tests always run with Docker and never with rkt?

As consequence, the test is skipped on all images that run node e2e tests. Is there a plan to move to cos-60-beta, where the test could run?

jsafrane commented:
/retest

kfox1111 commented:
#46643 raises an interesting question. Maybe hostPath isn't the only volume plugin that should support shared mounts. Its also useful for emptyDir. Think fuse file system in a sidecar and shared via emptyDir to the rest of its pod.

jsafrane commented:
/retest

croomes commented:
This works well with a StorageOS DaemonSet (https://gist.github.com/croomes/2019ed1f4178bd199bf7c65e6379dcaa).  Would be happy to help with other testing...

feiskyer commented on pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto:
```diff
@@ -158,6 +164,8 @@ message Mount {
     bool readonly = 3;
     // If set, the mount needs SELinux relabeling.
     bool selinux_relabel = 4;
+    // Requested propagation mode
```
nit: add `.` at end.

jsafrane commented:
/retest

jsafrane commented:
/retest

luxas commented:
ping @kubernetes/sig-node-pr-reviews 

@mtaufen Why aren't we using a feature gate for this?

dixudx commented:
/cc

jsafrane commented:
/retest

jsafrane commented:
Updated according to changes in https://github.com/kubernetes/community/pull/659

- Alpha feature flag is used instead of `--experimental-mount-propagation`
- `MountPropagation` field moved from `struct VolumeMount` (part of Pod definition) to `HostPathVolumeSource` (part of PersistentVolume or Pod)
- No alpha annotations

(The last commit with e2e_node test still needs some changes, will fix next week)

jsafrane commented:
I dropped the e2e test. The test needs MountPropagation=true feature gate but node e2e test runs internal API server without any FeatureGates. I will post a separate PR that will fix that and add MountPropagation test when this one is merged.

jsafrane commented:
/retest

jsafrane commented:
Updated according to latest version of https://github.com/kubernetes/community/pull/589

saad-ali commented on pkg/apis/settings/validation/validation.go:
```diff
@@ -47,7 +47,7 @@ func ValidatePodPresetSpec(spec *settings.PodPresetSpec, fldPath *field.Path) fi
 	allErrs = append(allErrs, apivalidation.ValidateEnv(spec.Env, fldPath.Child("env"))...)
 	allErrs = append(allErrs, apivalidation.ValidateEnvFrom(spec.EnvFrom, fldPath.Child("envFrom"))...)
-	allErrs = append(allErrs, apivalidation.ValidateVolumeMounts(spec.VolumeMounts, volumes, fldPath.Child("volumeMounts"))...)
+	allErrs = append(allErrs, apivalidation.ValidateVolumeMounts(spec.VolumeMounts, volumes, nil, fldPath.Child("volumeMounts"))...)
```
nit: put a comment e.g. `/* paramName */` next to nil so it is clear what is being set to `nil`.

saad-ali replied:
nit: For clarity, add a comment, e.g. `nil /* paramName */`, to indicate what parameter `nil` is being passed in for.

saad-ali reviewed (approved):
/lgtm
/approve

This looks good from storage-sig side. Thanks for all your work @jsafrane!

@kubernetes/sig-node-pr-reviews can we please get a review ASAP, we want to get this in and let it bake ASAP.

derekwaynecarr commented:
@dchen1107 @vishh -- did you guys have cycles to do the kubelet review for this, if not I will pick up.

vishh commented on api/openapi-spec/swagger.json:
```diff
@@ -62256,6 +62256,10 @@
       "type": "string"
      },
+     "mountPropagation": {
+      "description": "mountPropagation is the mode how are mounts in the volume propagated from the host to the container and from the container to the host. When not set, MountPropagationHostToContainer is used. This field is alpha in 1.8 and can be reworked or removed in a future release. Optional.",
```
nit: "mountPropagation determines how mounts are propagated from the host to container and the other way around"

jsafrane replied:
fixed

vishh replied:
It looks like the text hasn't been updated yet?

jsafrane replied:
It is now, sorry.

vishh commented on pkg/api/types.go:
```diff
@@ -1391,8 +1391,34 @@ type VolumeMount struct {
 	// Defaults to "" (volume's root).
 	// +optional
 	SubPath string
+	// mountPropagation is the mode how are mounts in the volume propagated from
```
nit: nit: "mountPropagation determines how mounts are propagated from the host to container and the other way around. When not set, defaults to MountPropagationHostToContainer."

jsafrane replied:
@bgrant0607 was against using world "default" when something does not fill the default in `default.go`. https://github.com/kubernetes/kubernetes/pull/19921#issuecomment-176575106

bgrant0607 replied:
Correct. Is there a reason we wouldn't fill in the default?

vishh commented on pkg/features/kube_features.go:
```diff
@@ -133,6 +133,12 @@ const (
+	// owner: @jsafrane
+	// alpha: v1.8
+	//
+	// Enable mount propagation of HostPath volumes.
```
Is it really tied to just HostPath volumes?

jsafrane replied:
fixed

vishh commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -188,12 +191,32 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 		}
 
+		var propagation runtimeapi.MountPropagation
+		if enableMountPropagation {
```
nit: Can this logic be in a separate method?

jsafrane replied:
extracted into a separate method

vishh reviewed (changes requested):
Just a few comments and nits. This is mostly ready to be merged!

jsafrane commented:
@vishh, thanks for your review. I fixed most of your comments, PTAL

vishh commented:
Only a single [nit](https://github.com/kubernetes/kubernetes/pull/46444/files#r135321885) is unresolved. This is good to go otherwise. I'll approve it as is.
@saad-ali and others wait till all the open comments are resolving before marking this pr with lgtm label.

/approve

jsafrane committed:
Add mount propagation to CRI protocol

CRI will blindly obey Kubelet decission about what propagation should be
used when.

jsafrane committed:
Implement mount propagation in kubelet

jsafrane committed:
Implement mount propagation in docker shim

jsafrane commented:
I got approval from sig-node and sig-storage, now the API
/assign @thockin @bgrant0607

jsafrane commented:
(I'll rebase when I get the approvals, it conflicts too fast)

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -1024,6 +1024,38 @@ func validatePathNoBacksteps(targetPath string, fldPath *field.Path) field.Error
+
+	privileged := container.SecurityContext != nil && container.SecurityContext.Privileged != nil && *container.SecurityContext.Privileged
+	if *mountPropagation == api.MountPropagationBidirectional && !privileged {
+		allErrs = append(allErrs, field.Invalid(fldPath, *mountPropagation, "Bidirectional mount propagation is available only to privileged containers"))
```
Should this be Forbidden ?

jsafrane replied:
fixed

thockin commented:
/approve

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -369,7 +402,8 @@ func (kl *Kubelet) GenerateRunContainerOptions(pod *v1.Pod, container *v1.Contai
 
-	opts.Mounts, err = makeMounts(pod, kl.getPodDir(pod.UID), container, hostname, hostDomainName, podIP, volumes)
+	mountPropagationEnabled := utilfeature.DefaultFeatureGate.Enabled(features.MountPropagation)
+	opts.Mounts, err = makeMounts(pod, kl.getPodDir(pod.UID), container, hostname, hostDomainName, podIP, volumes, mountPropagationEnabled)
```
Why passing `mountPropagationEnabled` multiple times when instead, you could check that directly in `translateMountPropagation()`?

jsafrane replied:
Good question. Initially there was `kubelet --experimental-mount-propagation` that needed passing all over the place and this is the last remainder. Removed.

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -207,6 +217,29 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+		return runtimeapi.MountPropagation_PROPAGATION_BIDIRECTIONAL, nil
+
+	default:
+		return 0, fmt.Errorf("Invalid MountPropagation mode: %q", mountMode)
```
s/Invalid/invalid
Error should start with a lower-case letter

jsafrane replied:
fixed

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -207,6 +217,29 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+	case mountMode == nil:
+		// HostToContainer is the default
+		return runtimeapi.MountPropagation_PROPAGATION_HOST_TO_CONTAINER, nil
+
```
remove the extra line.

jsafrane replied:
done

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -207,6 +217,29 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+
+	case *mountMode == v1.MountPropagationHostToContainer:
+		return runtimeapi.MountPropagation_PROPAGATION_HOST_TO_CONTAINER, nil
+
```
remove the extra line.

jsafrane replied:
done

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -207,6 +217,29 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+
+	case *mountMode == v1.MountPropagationBidirectional:
+		return runtimeapi.MountPropagation_PROPAGATION_BIDIRECTIONAL, nil
+
```
remove the extra line.

jsafrane replied:
done

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -188,12 +191,19 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+		if err != nil {
+			return nil, err
+		}
+		glog.V(5).Infof("Pod %s/%s container %s mount %s has propagation %s", pod.Namespace, pod.Name, container.Name, mount.Name, propagation)
```
use `format.Pod(pod)` to log pod-level information.

jsafrane replied:
done

jsafrane commented:
@yujuhong fixed your comments and rebased, PTAL

jsafrane commented:
/retest

yujuhong commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1547,8 +1547,34 @@ type VolumeMount struct {
+	// When not set, MountPropagationHostToContainer is used.
+	// This field is alpha in 1.8 and can be reworked or removed in a future
+	// release.
+	// Optional.
```
s/Optional/+optional

jsafrane replied:
ok

yujuhong commented on pkg/api/validation/validation.go:
```diff
@@ -1024,6 +1024,38 @@ func validatePathNoBacksteps(targetPath string, fldPath *field.Path) field.Error
 
+// validateMountPropagation verifies that MountPropagation field is valid and
+// allowed for given container.
+func validateMountPropagation(mountPropagation *api.MountPropagationMode, container *api.Container, fldPath *field.Path) field.ErrorList {
```
The api changes guideline for alpha fields also suggestss clearing the field before persisting to the storage. https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version

Not sure if it's needed but it's worth checking out.

jsafrane replied:
I did not find any feature gate checks in `pkg/registry/core`. Is really required?

jsafrane replied:
there is discussion about clearing alpha fields in https://github.com/kubernetes/kubernetes/pull/50924. I'll wait for result and fix mountPropagation field when necessary.

yujuhong commented on pkg/api/validation/validation.go:
```diff
@@ -1024,6 +1024,38 @@ func validatePathNoBacksteps(targetPath string, fldPath *field.Path) field.Error
+		return allErrs
+	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MountPropagation) {
+		allErrs = append(allErrs, field.Invalid(fldPath, mountPropagation, "mount propagation is disabled by feature-gate"))
```
The convention seems to be using `Forbidden`

jsafrane replied:
ok

yujuhong commented on pkg/kubelet/dockershim/helpers_test.go:
```diff
@@ -323,3 +323,77 @@ func TestMakePortsAndBindings(t *testing.T) {
+	}
+	result := generateMountBindings(mounts)
+
+	if len(result) != len(expectedResult) {
```
The test file uses assert/require. It'd be good to be consistent.
`assert.Equal(t, result, expectedResult, result)`

jsafrane replied:
ok

vishh commented:
/lgtm
/approve

yujuhong commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -207,6 +217,26 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
+		// mount propagation is disabled, use private as in the old versions
+		return runtimeapi.MountPropagation_PROPAGATION_PRIVATE, nil
+	}
+	switch {
```
If the feature is enabled, there is no way to specify `PRIVATE` and the default changes to HOST_TO_CONTAINER. 

This *might* be fine for Alpha, but can we actually change the default mode in Beta/GA? Sorry I didn't follow the PR proposal closely, so most likely this has been answered before.

jsafrane replied:
The proposal expects that HOST_TO_CONTAINER is the new default and PRIVATE is not needed at all. We need to test this theory in alpha.

yujuhong replied:
I'm slightly concerned that we won't gain enough confidence through feedback of this alpha feature. Changing the default behavior is always tricky.
I don't want to block the PR since the proposal has been through rounds of reviews. We can come back to this before graduating the feature to beta.

yujuhong commented:
Have a question and some nits. Holding the lgtm for now.

jsafrane commented:
@yujuhong fixed all your comments but clearing alpha fields, PTAL. Thanks!

yujuhong commented:
/lgtm

jsafrane commented:
/retest

hzxuzhonghu commented on pkg/api/validation/validation.go:
```diff
@@ -1024,6 +1024,38 @@ func validatePathNoBacksteps(targetPath string, fldPath *field.Path) field.Error
+		allErrs = append(allErrs, field.NotSupported(fldPath, *mountPropagation, supportedMountPropagations.List()))
+	}
+
+	if container == nil {
```
why no err returned about nil

jsafrane replied:
See the comment two lines below. Some `PodPreset` test was failing on it as it validated a PodSpec with no containers.

jsafrane replied:
... some PodPreset test was failing on it as it validated **VolumeMounts** with no containers.

hzxuzhonghu replied:
If container is previously  validated,  I do not think this `if container == nil `  make  sense

jsafrane committed:
Add API for mount propagation.

In fact, this is one annotation + its parsing & validation. Appropriate
kubelet logic that uses this annotation is in following patches.

jsafrane committed:
Regenerate API

jsafrane commented:
/retest

jsafrane commented:
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-e2e-gce-bazel

jsafrane commented:
/retest

jsafrane commented:
/test pull-kubernetes-e2e-gce-etcd3

jsafrane commented:
restoring lgtm after rebase of generated stuff and features.go

jsafrane committed:
Add feature gate for mount propagation

jsafrane commented:
rebased, restoring lgtm again

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/retest
